server/filestore.go

```diff
@@ -2136,6 +2141,76 @@ func (fs *fileStore) enforceBytesLimit() {
 	}
 }
 
+// Will make sure we have limits honored for max msgs per subject on recovery.
+// We will make sure to go through all msg blocks etc. but in practice this
+// will most likely only be the last one, so can take a more conservative approach.
+// Lock should be held.
+func (fs *fileStore) enforceMsgPerSubjectLimit() {
+	maxMsgsPer := uint64(fs.cfg.MaxMsgsPer)
+
+	// We want to suppress callbacks from remove during this process
+	// since these should have already been deleted and accounted for.
+	cb := fs.scb
+	fs.scb = nil
+	defer func() { fs.scb = cb }()
+
+	// Also suppress writeIndex info since we will do that manually at the end.
+	fip := fs.fip
+	fs.fip = false
```

Comment:
As seen in a test failure, this causes removeMsg() to spin a go routine to do the flush asynchronously, which was detected by a data race. Maybe you need a new boolean in fs struct to handle this case?

Reply:
Ah thanks, its just optimization and we have a flush timer to not go too fast, will remove.

Reply:
fixed